### ISSUE TITLE
Version Control with the new IDE

### ISSUE THREAD

**Issue by mwolfe02** [WaynePhillipsEA/twinbasic#823]
**Is your feature request related to a problem? Please describe.**
The twinBASIC VS Code extension is not being developed until at least 2023.  [My export to source batch file](https://nolongerset.com/version-control-with-twinbasic/#commento-a340181ba3552fe5c0b69f776bc9e91e6d8846c7f8da7b90423c0701984e1e32) currently relies on the VS Code extension for generating version-control-friendly text files from tB .twinproj files.

**Describe the solution you'd like**
Guidance on what the best approach to version control is under the new IDE.

**Describe alternatives you've considered**
Continue using the VS Code extension for version control purposes (while using the new IDE for development).

**Additional context**
The current process of downloading and unzipping the tB IDE .exe file means that there is no reliable way to identify the location of the latest version of the IDE, as I was able to do with the VSCode extension.

**Comment by WaynePhillipsEA** [comment]
For now, how about we offer import/export via menus and keyboard shortcuts?  Also let's allow you to preset the export path in the project settings so that you don't have to enter it each time.

We could also offer a switch for auto-export for when the project is saved.  This would automate the process as much as we can before later offering integrated version control support.

**Comment by GavinBl** [comment]
Wayne, your suggested im/export via menu and on Save, and preset Export Path all sound great, thank you.

I'm using your command line but found the \Sources\ file was named "HelloWorld.twin", despite the batch file being aimed at the GetIPAddresses project. Is that my mistake somehow?

C:\tB>SET twinproj_path="C:\tB\App\projects\Sample 2. GetIPAddresses\projectName.twinproj"
C:\tB>SET export_folder_path="C:\tB\Proj\Source"
C:\tB>C:\tB\App\bin\twinBASIC_win32.exe export "C:\tB\App\projects\Sample 2. GetIPAddresses\projectName.twinproj" "C:\tB\Proj\Source" --overwrite
twinBASIC v0.15.27 by Wayne Phillips exporting from "C:\tB\App\projects\Sample 2. GetIPAddresses\projectName.twinproj" to "C:\tB\Proj\Source"...
DONE: C:\tB\Proj\Source\Settings
DONE: C:\tB\Proj\Source\Sources\HelloWorld.twin
DONE: C:\tB\Proj\Source\Resources\ICON\twinBASIC.ico
DONE: C:\tB\Proj\Source\Resources\MANIFEST#1.xml
DONE: C:\tB\Proj\Source.meta ...
DONE
C:\tB>pause Press any key to continue . . .

**Comment by bclothier** [comment]
Just as a FYI. I had a moment of uncertainty with how to organize the structures for import/export. For a GitHub project, it's common to sprinkle the repository with various markdown files and licenses. However I did not have any idea how tB will react to non-source files such as `readme.md` or `license` so I ended up putting it in a subfolder. For an example of this, see [this GH project](https://github.com/bclothier/TwinBasicSevenZip) where we have a `TwinBasic` subfolder to house all source files from import/export to avoid mixing with the non-source files contained at the root folder. I would like to avoid the need to have a subfolders as having a `readme.md` may be helpful to view in the IDE.

**Comment by WaynePhillipsEA** [comment]
> Just as a FYI. I had a moment of uncertainty with how to organize the structures for import/export. For a GitHub project, it's common to sprinkle the repository with various markdown files and licenses. However I did not have any idea how tB will react to non-source files such as `readme.md` or `license` so I ended up putting it in a subfolder. For an example of this, see [this GH project](https://github.com/bclothier/TwinBasicSevenZip) where we have a `TwinBasic` subfolder to house all source files from import/export to avoid mixing with the non-source files contained at the root folder. I would like to avoid the need to have a subfolders as having a `readme.md` may be helpful to view in the IDE.

In the VS Code IDE, it allows you create LICENCE.md, README.md and CHANGELOG.md in the restricted root folder, which was primarily added for packages so that we could have some consistency.  It looks like the new IDE is not allowing those files to be created presently, so I will get that fixed.

**Comment by WaynePhillipsEA** [comment]
> I'm using your command line but found the \Sources\ file was named "HelloWorld.twin", despite the batch file being aimed at the GetIPAddresses project. Is that my mistake somehow?

No, that's a mistake in that sample.  If you look at the file explorer when you've got the project open in the IDE, you'll see that the source file is wrongly named `HelloWorld.twin`.  I'll get that fixed, thanks.

**Comment by bclothier** [comment]
> It looks like the new IDE is not allowing those files to be created presently, so I will get that fixed.

But keep in mind that GH (or git in general) allows you to add any files/folders to the root folder. For optimal experience, it should be possible to have those available even though it wouldn't be used as a source for the compiler?

**Comment by WaynePhillipsEA** [comment]
So quite a few bits added for this in BETA 30.   First of all, there's a manual Export option in the File menu (Ctrl+E shortcut):

<img width="201" alt="tbExportMenu" src="https://user-images.githubusercontent.com/25738109/166716982-5c46e2ba-64f9-4055-a4a5-de37270fc837.png">

If you want to have this command export to a fixed path every time, you can set a path in the project settings:

<img width="509" alt="tbExportFixedPath" src="https://user-images.githubusercontent.com/25738109/166717794-ab6bf4df-39ae-47c6-90ea-775caadd303a.png">

And there is an 'Export After Save' option:

<img width="509" alt="tbExportAfterSave" src="https://user-images.githubusercontent.com/25738109/166718047-91d01435-201a-4ee5-a16b-5e6cf3a0787d.png">

### With the above two options you can now automate having the twinproj file exported to a fixed path of your choosing, every time the project is saved.

<hr>

Furthermore, there is now a 'Import from Folder' option in the new-project dialog, allowing you to easily import a previously exported project into a new project:

<img width="392" alt="tbImportFromFolder" src="https://user-images.githubusercontent.com/25738109/166718534-23663344-28b3-4d0b-b1a3-b865cb62f32d.png">

**Comment by bclothier** [comment]
Nice. When exporting & sharing on GH, do you think the the `.meta` file ought to be ignored? It looks like it ought to be. 

If the export path is a part of the `Settings` file, that may pose a problem for two or more contributors working on the same project with a different export path on their system.

**Comment by WaynePhillipsEA** [comment]
> When exporting & sharing on GH, do you think the the `.meta` file ought to be ignored? It looks like it ought to be.

I forgot to mention in the release notes that the meta file will no longer be exported in BETA 30.

> If the export path is a part of the `Settings` file, that may pose a problem for two or more contributors working on the same project with a different export path on their system.

I'm planning to address this in tomorrow's update.  Thanks.

**Comment by WaynePhillipsEA** [comment]
> If the export path is a part of the `Settings` file, that may pose a problem for two or more contributors working on the same project with a different export path on their system.

Probably the simplest solution would be to allow relative paths here (relative to the twinproj file)?

**Comment by bclothier** [comment]
I think that working with a relative path to the project file is the most typical setup for various languages and is easier to work with. Absolute paths in VBP makes source code control significantly more painful and we want to avoid that. Per-machine settings also can get messy.

One more thing to consider, though:
```
\Sources '<-- the sub folder to use as target for export
myProject.twinProj '<-- the original twinproj project file..
Copy of myProject.twinProj '<-- where does the export go now?
```

I think that in the case of `Copy of myProject.twinProj`, it should still export to the `\Sources` (as per the existing settings when we make the copy of `myProject.twinProj` file) but not sure if others might expect a separate subfolder. Because we can have the `twinProj` as a the container, it is possible to just make a copy of the file as a backup, which now is a separate history from what's in the source control and that can create significant headaches when reconciling changes and require additional discipline on the developer's part to not flip between two different methods of making histories/backups. For this reason, I think that a copy of the file should still work with the same folder (from relative .path) and encouraging making a copy in a separate folder if they don't want to mix the history for some reason.

**Comment by WaynePhillipsEA** [comment]
I'd agree with that assessment @bclothier.  What I might do, is use placeholders, similar to what we do for the `Project: Build Output Path` setting.  So you'll set it something like this:

`${SourcePath}\${ProjectName}\`

or if you want to include the actual project filename (to disambiguate in the example you gave):

`${SourcePath}\${ProjectFileName}\`

That will allow some flexibility.

**Comment by EduardoVB** [comment]
> Absolute paths in VBP makes source code control significantly more painful

AFAI remember VB6 works with relative paths. Maybe somewhere there were absolute paths?

When the files are in another drive of course they need to be addressed by absolute paths, maybe that can cause confusion.
